processor/src/chiplets/memory/segment.rs

```diff
@@ -72,37 +76,60 @@ impl MemorySegmentTrace {
     ///
     /// If the specified address hasn't been previously written to, four ZERO elements are
     /// returned. This effectively implies that memory is initialized to ZERO.
-    pub fn read(&mut self, addr: u32, clk: Felt) -> Word {
+    pub fn read(&mut self, ctx: ContextId, addr: u32, clk: Felt) -> Result<Word, ExecutionError> {
         // look up the previous value in the appropriate address trace and add (clk, prev_value)
         // to it; if this is the first time we access this address, create address trace for it
         // with entry (clk, [ZERO, 4]). in both cases, return the last value in the address trace.
-        self.0
-            .entry(addr)
-            .and_modify(|addr_trace| {
-                let last_value = addr_trace.last().expect("empty address trace").value();
-                let access = MemorySegmentAccess::new(clk, MemoryOperation::CopyRead, last_value);
-                addr_trace.push(access);
-            })
-            .or_insert_with(|| {
+        match self.0.entry(addr) {
+            Entry::Vacant(vacant_entry) => {
                 let access =
                     MemorySegmentAccess::new(clk, MemoryOperation::InitRead, INIT_MEM_VALUE);
-                vec![access]
-            })
-            .last()
-            .expect("empty address trace")
-            .value()
+                vacant_entry.insert(vec![access]);
+                Ok(INIT_MEM_VALUE)
+            },
+            Entry::Occupied(mut occupied_entry) => {
+                let addr_trace = occupied_entry.get_mut();
+                if addr_trace.last().expect("empty address trace").clk() == clk {
+                    Err(ExecutionError::DuplicateMemoryAccess { ctx, addr, clk })
+                } else {
+                    let last_value = addr_trace.last().expect("empty address trace").value();
+                    let access =
+                        MemorySegmentAccess::new(clk, MemoryOperation::CopyRead, last_value);
+                    addr_trace.push(access);
+
+                    Ok(last_value)
+                }
+            },
+        }
     }
 
     /// Writes the provided word at the specified address. The memory access is assumed to happen
     /// at the provided clock cycle.
-    pub fn write(&mut self, addr: u32, clk: Felt, value: Word) {
+    pub fn write(
+        &mut self,
+        ctx: ContextId,
+        addr: u32,
+        clk: Felt,
+        value: Word,
+    ) -> Result<(), ExecutionError> {
```

Comment:
Similar to the last comment, I'd update doc comments to explain under what conditions an error could happen.